estat/README.md

```diff
@@ -0,0 +1,156 @@
+
+
+                                       iops(/s)  throughput(k/s)
+ total                                       42              765
+
+```
+
+##  Performance data collector structure
+
+The main `cmd/estat` script adds the needed boiler plate for:  
+
+- `BCC` front-end (loading the eBPF program, attaching probes etc).
+- Common I/O data aggregation with: `AGGREGATE_DATA()` and `aggregate1(), aggregate2(), aggregate3()` depending on options (`lzqy`) passed
+- Helper (`bcchelper`) for various aggregation methods (sum, average, log, log linear histograms etc) and printing formatted output.  
+
+An `estat` performance data collector or program then, has the below structure leveraging the framework mentioned above  
+
+- Defines a base hash map for collecting data  
+    For example:
+
+    ```
+    // Structure to hold thread local data
+    typedef struct {
+       u64 ts;
+    } zio_data_t;
+    BPF_HASH(zio_base_data, zio_t *, zio_data_t);
+    ```
+
+- Uses kprobes (`kprobe`, `kretprobe`) for dynamic tracing and `AGGREGATE_DATA()` for any I/O aggregation  
+    For example:
+
+    ```
+    // @@ kretprobe|vdev_queue_io_to_issue|vdev_queue_issue
+    int vdev_queue_issue(struct pt_regs *ctx)
+    {
+         zio_data_t data = {};
+         data.ts = bpf_ktime_get_ns();
+     
+         if (zio == NULL)
+            return 0;
+     
+         zio_base_data.update(&zio, &data);
+     
+         return 0;
+    }
+     
+     // @@ kprobe|vdev_queue_io_done|vdev_queue_done
+    int vdev_queue_done(struct pt_regs *ctx, zio_t *zio)
+    {
+        u64 ts = bpf_ktime_get_ns();
+        zio_data_t *data = zio_base_data.lookup(&zio);
+        if (data == 0) {
+            return 0;   // missed issue
+        }
+    
+        u64 delta = ts - data->ts;
+        char name[NAME_LENGTH];
+        char axis[AXIS_LENGTH];
+    
+        if (zio->io_type == 1) {
+            __builtin_memcpy(&name, "read ", OP_LENGTH + 1);
+        } else if (zio->io_type == 2) {
+            __builtin_memcpy(&name, "write ", OP_LENGTH + 1);
+        }
+       if (zio->io_priority == 0) {
+           __builtin_memcpy(&axis, "syncr", PRIORITY_LENGTH);
+       } else if (zio->io_priority == 1) {
+           __builtin_memcpy(&axis, "syncw", PRIORITY_LENGTH);
+       }
+    
+       AGGREGATE_DATA(name, axis, delta, zio->io_size);
+       zio_base_data.delete(&zio);
+    
+       return 0;
+    }
+    ```
+
+- The statements starting with `// @@ ` (example above) **are required** for the probe functions to be attached via a pyton BPF object.  
+  The statement specifies the text necessary to complete an attach call when appened to `b.attach_`.  
+  For example:
+
+      `// @@ kprobe|vdev_queue_io_done|vdev_queue_done`
+
+    leads to this attach call:
+
+      `b.attach_kprobe(event="vdev_queue_io_done", fn_name="vdev_queue_done")`
+
+- The `name` and optionally an `axis` define an `I/O type` for which aggregations are done and output.
```

Comment:
> The `name` and optionally an `axis`

I find the terminology here a bit strange, especially `axis`. I wonder if it would be clearer to call these something like `io_type` and `io_subtype`. Not very high priority, so if we do want to change this, let's do it after the initial PR is merged.

Reply:
Ok lets discuss offline about this.